### PR TITLE
CONTRIBUTING.md: add section on testing

### PR DESCRIPTION
**Pull request by skriss**
Updates #3312.

Signed-off-by: Steve Kriss <krisss@vmware.com>

### CODE REVIEW

**Comment by skriss** [comment]
I wrote up some notes for myself and thought these would be useful to add to the docs too.

**Comment by skriss** [comment]
> # [Codecov](https://codecov.io/gh/projectcontour/contour/pull/3434?src=pr&el=h1) Report
> > Merging [#3434](https://codecov.io/gh/projectcontour/contour/pull/3434?src=pr&el=desc) ([2994879](https://github.com/projectcontour/contour/commit/299487974cc58276d3efc7b10bff12606746a575)) into [main](https://codecov.io/gh/projectcontour/contour/commit/885e28d0f7ad744d81fd2b3e1c105ec4689472cb?el=desc) ([885e28d](https://github.com/projectcontour/contour/commit/885e28d0f7ad744d81fd2b3e1c105ec4689472cb)) will **decrease** coverage by `0.03%`.
> > The diff coverage is `n/a`.
> 
> [Impacted file tree graph](https://codecov.io/gh/projectcontour/contour/pull/3434?src=pr&el=tree)
> 
> ```diff
> @@            Coverage Diff             @@
> ##             main    #3434      +/-   ##
> ==========================================
> - Coverage   75.03%   75.00%   -0.04%     
> ==========================================
>   Files          98       98              
>   Lines        6600     6600              
> ==========================================
> - Hits         4952     4950       -2     
> - Misses       1536     1537       +1     
> - Partials      112      113       +1     
> ```
> 
> [Impacted Files](https://codecov.io/gh/projectcontour/contour/pull/3434?src=pr&el=tree)	Coverage Δ	
> [internal/dag/cache.go](https://codecov.io/gh/projectcontour/contour/pull/3434/diff?src=pr&el=tree#diff-aW50ZXJuYWwvZGFnL2NhY2hlLmdv)	`91.15% <0.00%> (-0.59%)`	⬇️

Huh? 😂

**Comment by sunjayBhatia** [comment]
> > # [Codecov](https://codecov.io/gh/projectcontour/contour/pull/3434?src=pr&el=h1) Report
> > > Merging [#3434](https://codecov.io/gh/projectcontour/contour/pull/3434?src=pr&el=desc) ([2994879](https://github.com/projectcontour/contour/commit/299487974cc58276d3efc7b10bff12606746a575)) into [main](https://codecov.io/gh/projectcontour/contour/commit/885e28d0f7ad744d81fd2b3e1c105ec4689472cb?el=desc) ([885e28d](https://github.com/projectcontour/contour/commit/885e28d0f7ad744d81fd2b3e1c105ec4689472cb)) will **decrease** coverage by `0.03%`.
> > > The diff coverage is `n/a`.
> > 
> > 
> > [Impacted file tree graph](https://codecov.io/gh/projectcontour/contour/pull/3434?src=pr&el=tree)
> > ```diff
> > @@            Coverage Diff             @@
> > ##             main    #3434      +/-   ##
> > ==========================================
> > - Coverage   75.03%   75.00%   -0.04%     
> > ==========================================
> >   Files          98       98              
> >   Lines        6600     6600              
> > ==========================================
> > - Hits         4952     4950       -2     
> > - Misses       1536     1537       +1     
> > - Partials      112      113       +1     
> > ```
> > 
> > 
> > [Impacted Files](https://codecov.io/gh/projectcontour/contour/pull/3434?src=pr&el=tree)	Coverage Δ
> > [internal/dag/cache.go](https://codecov.io/gh/projectcontour/contour/pull/3434/diff?src=pr&el=tree#diff-aW50ZXJuYWwvZGFnL2NhY2hlLmdv)	`91.15% <0.00%> (-0.59%)`	⬇️
> 
> Huh? 😂

There must be some test that is non deterministic? Or the coverage measurement is flaky

https://codecov.io/gh/projectcontour/contour/pull/3434/changes seems to be the coverage difference?

**Review by sunjayBhatia — APPROVED** [review]
Nice! I like this addition

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -191,6 +191,37 @@ This will push to `:latest` in `docker.io/davecheney` obviously you'll also need
...
 To verify your change by deploying the image you built, take one of the [deployment manifests][7], edit it to point to your new image, and deploy to your Kubernetes cluster.
 
+## Contour testing
+
+This section provides some useful information and guidelines for working with Contour's tests.
+
+### Glossary
+
+#### Config/Data Categories
+* **Kubernetes Config**: `HTTPProxy`, `Ingress` or `Gateway` config that Contour watches and converts to Envoy config.
```
Comment (stevesloka) [inline CONTRIBUTING.md:201]: Maybe call this `GatewayAPI` or `gateway-api`?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -191,6 +191,37 @@ This will push to `:latest` in `docker.io/davecheney` obviously you'll also need
...
+* **DAG**: The internal Contour representation of L7 proxy concepts. Kubernetes config is first converted to DAG objects before being converted to Envoy config.
+* **Envoy Config**: Configuration that can be provided to Envoy via xDS. This is Contour's final output, generated directly from the DAG. 
+
+#### Test Categories
+* **Unit Test**: A Go test for a particular function/package. In some cases, these test more than one package at a time.
+* **Feature Test**: A Go test in `internal/featuretests` that tests the translation of Kubernetes config to Envoy config, using a Contour event handler and xDS server. 
+* **Integration Test**: A YAML/rego test in `_integration/testsuite` that performs a full end-to-end test of Contour running in a cluster. Typically verifies the behavior of HTTP requests given an `HTTPProxy` config.
```
Comment (stevesloka) [inline CONTRIBUTING.md:208]: nit: We do have tests (well a single) for GatewayAPI.

**Comment by skriss** [comment]
@stevesloka comments addressed.

**Review by youngnick — APPROVED** [review]

**Pull request merged**